sensor/common/registry/registry_store.go

```diff
@@ -392,3 +393,220 @@ func (rs *Store) GetMatchingCentralRegistryIntegrations(imgName *storage.ImageNa
 
 	return regs
 }
+
+// UpsertSecret upserts a pull secret into the store.
+func (rs *Store) UpsertSecret(namespace, secretName string, dockerConfig config.DockerConfig, serviceAcctName string) {
+	if !features.SensorPullSecretsByName.Enabled() {
+		rs.upsertSecretByHost(namespace, secretName, dockerConfig, serviceAcctName)
+		return
+	}
+
+	rs.upsertSecretByName(namespace, secretName, dockerConfig, serviceAcctName)
+}
+
+func (rs *Store) upsertSecretByHost(namespace, secretName string, dockerConfig config.DockerConfig, serviceAcctName string) {
+	isGlobalPullSecret := namespace == openshiftConfigNamespace && secretName == openshiftConfigPullSecret
+	fromDefaultSA := serviceAcctName == defaultSA
+
+	for registryAddress, dce := range dockerConfig {
+		registryAddr := strings.TrimSpace(registryAddress)
+
+		if fromDefaultSA {
+			// We assume that registries found in the dockercfg secret managed by OCP for the default
+			// service account only references hostnames for the OCP internal registry.
+			rs.addClusterLocalRegistryHost(registryAddr)
+			if err := rs.upsertRegistry(namespace, registryAddr, dce); err != nil {
+				log.Errorf("Unable to upsert registry %q into store: %v", registryAddr, err)
+			}
+			continue
+		}
+
+		if env.DelegatedScanningDisabled.BooleanSetting() {
+			// If delegated scanning is disabled then we do not store additional secrets outside of those needed
+			// for scanning images from the OCP internal registry.
+			continue
+		}
+
+		if serviceAcctName != "" {
+			// Ignore secrets for service accounts other than default so that the
+			// default registry is not overwritten in the store.
+			continue
+		}
+
+		var err error
+		if isGlobalPullSecret {
+			err = rs.upsertGlobalRegistry(registryAddr, dce)
+		} else {
+			err = rs.upsertRegistry(namespace, registryAddr, dce)
+		}
+		if err != nil {
+			log.Errorf("unable to upsert registry %q into store: %v", registryAddr, err)
+		}
+	}
+}
+
+func (rs *Store) upsertSecretByName(namespace, secretName string, dockerConfig config.DockerConfig, serviceAcctName string) {
+	isGlobalPullSecret := namespace == openshiftConfigNamespace && secretName == openshiftConfigPullSecret
+
+	// To avoid partial upserts - hold the lock until the entire secret upserted.
+	rs.storeMutux.Lock()
+	defer rs.storeMutux.Unlock()
+
+	for registryAddress, dce := range dockerConfig {
+		registryAddr := strings.TrimSpace(registryAddress)
+
+		if serviceAcctName != "" {
```

Comment:
related to my comment above. I was confused about the comments you left right below this line, but I now I realize this means that OpenShift Internal Registry creds are taken from those secrets with service account annotations. I looked at the original implementation, and the creds were only taken from the secret annotated with the default service account. I understand we now want to account for the secret's name, but is that necessary for the OpenShift Internal Registry?

Reply:
I debated this when creating the PR - TL;DR; it may not be 'necessary' - with what we know now we will likely get successful scans by assuming secrets associated with the `default` service account will work for any image in the OCP registry for that namespace.  OCP however will use the specific secrets associated with the workloads specific service account, and I believe we should too to future proof this.  It also removes some complexity by not having special treatment with regards to pulling the correct secrets (with the tradeoff being more secrets are stored).

As a hypothetical (not sure if currently possible) - if access to images within the OCP internal registry can be restricted to specific service accounts - then ACS would just work by not assuming special treatment.

If you feel strongly about this, can add the special treatment, perhaps behind an env var - the benefit being less secrets stored, the caveats being extra logic for the special treatment and not aligned with what OCP does.